### PR TITLE
fix Ramda ifElse to work with variadic arguments

### PR DESCRIPTION
**Pull request by LoganBarnett**
- https://ramdajs.com/docs/#ifElse
- #2411
- Type of contribution: fix

Other notes:

This fixes the issue found in #2411 where `test` could not be used with `ifElse` because `test` is automatically curried, which relies upon variadic arguments.

As another note, I noticed as part of working on this that some of our tests have a lot of cross-dependencies. For example, the `ifElse` test also depended on `is`. I worry that if something happens with `is`, the test for `ifElse` can become invalid. Perhaps instead we can write our tests by only using a single Ramda function at a time, so our test doesn't potentially become a lie built atop another lie. For example, if `is` returns or takes `any` it could invalidate the `ifElse` test entirely. It also conflates what's really being tested. In the case of `test` from #2411, if it dropped variadic argument support while the `ifElse` test I used actually used `test`, then the test would silently become invalid yet pass.

TLDR; Going forward, let's use only one Ramda function at a time during an assertion.

Thoughts?

From the commit message:

This is very typical when working with curried functions, particularly from within Ramda itself. The prior definition of ifElse used to work in type refinement scenarios (although really, it appears to be coercing to any or otherwise losing type information).

The original test for ifElse is broken because it relies upon such refinements or loss of type information.

The tests for ifElse are now expressed in a way that is independent of other Ramda functions.

### CODE REVIEW

File: definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js [hunk 1]
```diff
@@ -61,11 +72,58 @@ const def1: number = defaultTo42(undefined);
...
+    // $ExpectError
+    const result: void | string = ifElseFn(5)
+  })
+
+  // This was the original test for ifElse, which depends on the implementation
+  // of "is". "is" in its present form does not perform type refinement, so the
+  // stricter form of ifElse fails the check with this test. This test below
+  // captures the issue, and is left for reference and/or if some brave soul
+  // wishes to expand the definition of ifElse (and possibly also "is") to
+  // refine the predicate correctly. At time of writing, $Pred and $Refine are
```
Comment (AndrewSouthpaw) [inline definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js]: > and/or if some brave soul wishes to expand the definition

😂

File: definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js [hunk 2]
```diff
@@ -61,11 +72,58 @@ const def1: number = defaultTo42(undefined);
...
+    // Because the object literal ({}) isn't a number it
+    // $ExpectError
+    const ie: number | string = incCount({});
+  })
+
+  it('works with variadic arg inputs', () => {
+    const ifElseFn = ifElse(
+      // The original motivation was to work with "test". Future changes could
+      // invalidate the nature of the test, so we need to mimic the form of test
+      // at time of writing. The error will be along the lines of this:
```
Comment (AndrewSouthpaw) [inline definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js]: I'm a little confused because we're writing a test about `test` in the context of `ifElse`. Would you mind expanding a bit here and disambiguating? I'm still having trouble understanding this point.

**Comment by AndrewSouthpaw** [comment]
> TLDR; Going forward, let's use only one Ramda function at a time during an assertion.

Agreed!

**Comment by AndrewSouthpaw** [comment]
Also... considering your mention of [nmn's solution about variadic generics](https://github.com/facebook/flow/issues/1251#issuecomment-289924710) I'm a little surprised you didn't need to provide a `$ReadOnlyArray<mixed>`.

File: definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js (hunk 2 above)
Comment (LoganBarnett) [reply definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js]: Ah - this is perhaps the wrong kind of context I'm providing. The original ticket this is from is essentially "`test` does not work with `ifElse`", when really the problem is `ifElse` doesn't work with varargs. This might have worked as a self-made PR comment instead of a comment in the code, since the ticket itself and a PR comment are ephemeral to each other compared to the code that will outlast the others' relevances.

I'll remove this comment.

**Comment by LoganBarnett** [comment]
I remember taking out `$ReadOnlyArray<mixed>` because of existing tests that broke from it. It might have been a red herring though, because of the `is` call kind of falsely doing a refinement. My memory isn't the most trustworthy thing though. I tried to reproduce the issue with `$ReadOnlyArray<mixed>` in [this try](https://flow.org/try/#0PTAEBUAsEsGdQC4FNYPgQQE4HN4GMB7AO1UwENoikATRA0AEgCUkzqB5IgGwE8tyeAHgC20AB40AfACgQESElAAHTAQBGSLsNBlMsOClAB3BUVA8CAVx0JkwpQjqhLsRQgXLdZYUmR7QlIgKsmAImJaKZES0AGZkXK6geGSu8IHuimrkRHiQoDGWOQjQxPAAtGVBSDw6mIr69rwhoKIStJSorLQEMcp1rpgAbjQAdNLS1Eh4XLqKBUUlZtAxAKIJSIJYuABcjCxsnLz8ZEKtUgA0oABClwDCkgAU0qAvScTUuw8j37o7oFuwACUoAAvJJQGoCAQuKwiM9QIDPk9XvkAIyfb4jX6wXYA4Fg67w-Hgh4xABMGJ+OBx-2pxNAt3pXypfzxoPBV1AAB8GcCAGTwl4PZEotGUrHU3F09nXc6C17k8XYqW4em3eVMzHK2mqmWcnmMgDc8Mm01m+UKeGKxACq3Wm0ley6hz4mAEInEF1lDMe8sI0SVjrZBMh0NhcpRMXRoGZEtZ0oJN3lipjWqDCfB6peiNTLJpwY53IZxuk-tQb2ioNta1cItj2oA2qRKNgALo50MwqIysIRCMxmnNojYemwEYIAgAVSUSiQmFuKSQD0B-Yeg7CLdH44IABkCEY5wvayvpICS2XHP1LFwELtBgRoLQeUPsFX-dQHgByShKSwIT+As0DArGIs5WisboEJgIQXqAV43neD5PqAL5vu8DwAKyAaWpSOJQeC3FYRCOCC1brA8YgypRACkoBkqCpEAAyXJRBKUQA1KAqIsVR24AMobsOy5njhJB4UguxEJYwgaJgRaoaR+GEYUCAPKiIlyMBoFTAgEGqNBciwdAEmgFJMlzvJgmvopOTKcRDwAN4AL5nkAA) but I cannot any longer. Both `Args` and `Args: $ReadOnlyArray<mixed>` work here. Do you have a preference on which you'd like to see?

File: definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js (hunk 2 above)
Comment (LoganBarnett) [reply definitions/npm/ramda_v0.x.x/flow_v0.82.x-/test_ramda_v0.x.x_logic.js]: Okay this should be addressed now. There is a comment there still, but the `test` context is now gone.

**Comment by LoganBarnett** [comment]
I forgot to apply the changes to the latest Ramda (0.26.x). It's just a straight copy-pasta.

**Comment by AndrewSouthpaw** [comment]
>  It's just a straight copy-pasta.

https://www.youtube.com/watch?v=cE1FrqheQNI
